core/shared/src/main/scala/scalaz/zio/IO.scala

```diff
@@ -1030,6 +1030,45 @@ object IO extends Serializable {
    */
   final def halt[E](cause: Cause[E]): IO[E, Nothing] = new Fail(cause)
 
+  /**
+   * Imports a synchronous effect that does blocking IO into a pure value.
+   *
+   * If the returned `IO` is interrupted, the blocked thread running the synchronous effect
+   * will be interrupted via `Thread.interrupt`.
+   */
+  final def blocking[A](effect: => A): IO[Nothing, A] =
+    IO.flatten(IO.sync {
+      import java.util.concurrent.locks.ReentrantLock
+      import java.util.concurrent.atomic.AtomicReference
+
+      val lock   = new ReentrantLock()
+      val thread = new AtomicReference[Option[Thread]](None)
+
+      def withLock[B](b: => B): B =
+        try {
+          lock.lock(); b
+        } finally lock.unlock()
+
+      for {
+        finalizer <- Ref[IO[Nothing, Unit]](IO.unit)
+        a <- (for {
+              fiber <- (IO.unyielding(IO.sync[Either[Throwable, A]] {
+                        withLock(thread.set(Some(Thread.currentThread())))
```

Comment:
No, the thread will, after finishing this work, start working on some other fiber's work, and you don't want to accidentally kill it, so you need to make sure the thread stays working on the target fiber (rather than finishing and switching to another fiber) before you call Thread.interrupt. Or else you interrupt the wrong fiber.